packages/react/src/components/button.tsx

```diff
@@ -106,128 +104,104 @@ export function UploadButton<
-  const [isManualTriggerDisplayed, setIsManualTriggerDisplayed] =
-    useState(false);
-  const uploadProgress =
-    $props.__internal_upload_progress ?? uploadProgressState;
 
-  const { startUpload, isUploading, permittedFileInfo } = useUploadThing(
-    $props.endpoint,
-    {
-      headers: $props.headers,
-      skipPolling: !$props?.onClientUploadComplete ? true : $props?.skipPolling,
-      onClientUploadComplete: (res) => {
-        if (fileInputRef.current) {
-          fileInputRef.current.value = "";
-        }
-        setIsManualTriggerDisplayed(false);
-        setFiles([]);
-        $props.onClientUploadComplete?.(res);
-        setUploadProgress(0);
-      },
-      onUploadProgress: (p) => {
-        setUploadProgress(p);
-        $props.onUploadProgress?.(p);
-      },
-      onUploadError: $props.onUploadError,
-      onUploadBegin: $props.onUploadBegin,
-      onBeforeUploadBegin: $props.onBeforeUploadBegin,
+  const {
+    files,
+    setFiles,
+    startUpload,
+    isUploading,
+    permittedFileInfo,
+    getInputProps,
+  } = useUploadThing($props.endpoint, {
+    headers: $props.headers,
+    skipPolling: !$props?.onClientUploadComplete ? true : $props?.skipPolling,
+    onClientUploadComplete: (res) => {
+      if (fileInputRef.current) {
+        fileInputRef.current.value = "";
+      }
+      $props.onClientUploadComplete?.(res);
+      setUploadProgress(0);
     },
-  );
+    onUploadProgress: (p) => {
+      setUploadProgress(p);
+      $props.onUploadProgress?.(p);
+    },
+    onUploadError: $props.onUploadError,
+    onUploadBegin: $props.onUploadBegin,
+    onBeforeUploadBegin: $props.onBeforeUploadBegin,
+  });
 
-  const { fileTypes, multiple } = generatePermittedFileTypes(
-    permittedFileInfo?.config,
+  const inputProps = useMemo(
+    () => getInputProps({ mode }),
+    [getInputProps, mode],
   );
+  if ($props.__internal_button_disabled) inputProps.disabled = true;
 
-  const ready =
-    $props.__internal_ready ??
-    ($props.__internal_state === "ready" || fileTypes.length > 0);
+  const state = (() => {
+    if ($props.__internal_state) return $props.__internal_state;
+    if (inputProps.disabled) return "readying";
+    if (!inputProps.disabled && !isUploading) return "ready";
+    return "uploading";
+  })();
 
-  useEffect(() => {
-    const handlePaste = (event: ClipboardEvent) => {
-      if (!appendOnPaste) return;
-      if (document.activeElement !== labelRef.current) return;
+  usePaste((event) => {
+    if (!appendOnPaste) return;
+    if (document.activeElement !== fileInputRef.current) return;
 
-      const pastedFiles = getFilesFromClipboardEvent(event);
-      if (!pastedFiles) return;
+    const pastedFiles = getFilesFromClipboardEvent(event);
+    if (!pastedFiles) return;
 
-      setFiles((prev) => [...prev, ...pastedFiles]);
+    let filesToUpload = pastedFiles;
+    setFiles((prev) => {
+      filesToUpload = [...prev, ...pastedFiles];
+      return filesToUpload;
+    });
```

Comment:
This bug was fixed for dropzone in an earlier PR (https://github.com/pingdotgg/uploadthing/pull/622/files)